tests/catch/unit/stream/hipStreamDestroy.cc

```diff
@@ -0,0 +1,131 @@
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+#include <hip_test_common.hh>
+
+namespace hipStreamDestroyTests {
+
+TEST_CASE("Unit_hipStreamDestroy_Default") {
+  hipStream_t stream{};
+  HIP_CHECK(hipStreamCreate(&stream));
+  HIP_CHECK(hipStreamDestroy(stream));
+}
+
+TEST_CASE("Unit_hipStreamDestroy_DoubleDestroy") {
+  hipStream_t stream{};
+  HIP_CHECK(hipStreamCreate(&stream));
+  HIP_CHECK(hipStreamDestroy(stream));
+  HIP_CHECK_ERROR(hipStreamDestroy(stream), hipErrorInvalidResourceHandle);
+}
+
+TEST_CASE("Unit_hipStreamDestroy_NullStream") {
+  hipStream_t stream{};
+  HIP_CHECK_ERROR(hipStreamDestroy(stream), hipErrorInvalidResourceHandle);
+}
+
+__global__ void kernel(int* x, size_t size) {
+  size_t tid{threadIdx.x};
+  if (tid < size) {
+    x[tid] = 1;
+  }
+}
+
+TEST_CASE("Unit_hipStreamDestroy_WithFinishedWork") {
+  hipStream_t stream{};
+  HIP_CHECK(hipStreamCreate(&stream));
+
+  constexpr int data_points = 10;
+  int* device_data{};
+  HIP_CHECK(hipMalloc(&device_data, sizeof(int) * data_points));
+
+  kernel<<<1, data_points, 0, stream>>>(device_data, data_points);
+  HIP_CHECK(hipStreamSynchronize(stream));
+
+  std::array<int, data_points> host_data{};
+  HIP_CHECK(
+      hipMemcpy(host_data.data(), device_data, sizeof(int) * data_points, hipMemcpyDeviceToHost));
+
+  REQUIRE(std::all_of(std::begin(host_data), std::end(host_data), [](int x) { return x == 1; }));
+
+  HIP_CHECK(hipStreamDestroy(stream));
+  HIP_CHECK(hipFree(device_data));
+}
+
+/**
+ * @brief Kernel that signals a semaphore to change value from 0 to 1.
+ *
+ * @param semaphore the semaphore that needs to be signaled.
+ */
+__global__ void signaling_kernel(int* semaphore) {
+  size_t tid = blockIdx.x * blockDim.x + threadIdx.x;
+  if (tid == 0) {
+    atomicAdd(semaphore, 1);
+  }
+}
+
+/**
+ * @brief Kernel that busy waits until the specified semaphore goes from 0 to 1.
+ *
+ * @param semaphore the semaphore to wait for.
+ */
+__global__ void waiting_kernel(int* semaphore) {
+  size_t tid = blockIdx.x * blockDim.x + threadIdx.x;
+  if (tid == 0) {
+    while (atomicCAS(semaphore, 1, 2) == 0) {
+    }
+  }
+}
+
+// NOTE possible integration test
+TEST_CASE("Unit_hipStreamDestroy_WithPendingWork") {
```

Comment:
I don't like using timers like that as the system could be busy doing something else and all the assumed delays are off.
I don't feel like you need to add the thread to delay the completion of the kernel. You can prob do a spin loop inside the kernel or do the signaling from the host using hipHostAlloc memory instead of hipMalloc or something.